src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -40,131 +42,207 @@
+    @ForceInline
+    private static String readInt(AbstractMemorySegmentImpl segment, long offset, Charset charset) {
+        int len = strlenInt(segment, offset, segment.byteSize());
+        byte[] bytes = new byte[len];
+        MemorySegment.copy(segment, JAVA_BYTE, offset, bytes, 0, len);
         return new String(bytes, charset);
     }
 
-    private static void writeInt(MemorySegment segment, long offset, Charset charset, String string) {
+    @ForceInline
+    private static void writeInt(AbstractMemorySegmentImpl segment, long offset, Charset charset, String string) {
         int bytes = copyBytes(string, segment, charset, offset);
         segment.set(JAVA_INT_UNALIGNED, offset + bytes, 0);
     }
 
     /**
-     * {@return the shortest distance beginning at the provided {@code start}
-     *  to the encountering of a zero byte in the provided {@code segment}}
+     * {@return the shortest distance beginning at the provided {@code fromOffset}
+     *          to the encountering of a zero byte in the provided {@code segment}
+     *          checking bytes before the {@code toOffset}}
      * <p>
-     * The method divides the region of interest into three distinct regions:
-     * <ul>
-     *     <li>head (access made on a byte-by-byte basis) (if any)</li>
-     *     <li>body (access made with eight bytes at a time at physically 64-bit-aligned memory) (if any)</li>
-     *     <li>tail (access made on a byte-by-byte basis) (if any)</li>
-     * </ul>
-     * <p>
-     * The body is using a heuristic method to determine if a long word
-     * contains a zero byte. The method might have false positives but
-     * never false negatives.
+     * The method is using a heuristic method to determine if a long word contains a
+     * zero byte. The method might have false positives but never false negatives.
      * <p>
      * This method is inspired by the `glibc/string/strlen.c` implementation
      *
-     * @param segment to examine
-     * @param start   from where examination shall begin
+     * @param segment    to examine
+     * @param fromOffset from where examination shall begin (inclusive)
+     * @param toOffset   to where examination shall end (exclusive)
      * @throws IllegalArgumentException if the examined region contains no zero bytes
      *                                  within a length that can be accepted by a String
      */
-    public static int chunkedStrlenByte(MemorySegment segment, long start) {
-
-        // Handle the first unaligned "head" bytes separately
-        int headCount = (int)SharedUtils.remainsToAlignment(segment.address() + start, Long.BYTES);
-
-        int offset = 0;
-        for (; offset < headCount; offset++) {
-            byte curr = segment.get(JAVA_BYTE, start + offset);
-            if (curr == 0) {
-                return offset;
+    @ForceInline
+    public static int strlenByte(final AbstractMemorySegmentImpl segment,
```

Comment:
Note: clients of this method seems to always pass `segment.byteSize()` for `toOffset`. What is the value of passing it as an explicit parameter instead of having `byteSize()` being called inside the method?

Reply:
As this is an internal API, I tried to keep it as similar as possible to the other bulk operations. But as you say, the value used here for the last parameter is always `segment.byteSize()`. We could easily change it. What's your preference here?

Reply:
If the method works correctly on all end offsets (even when the offset is different from the segment size), then I'm ok leaving it as is. I'm mostly trying to make sure that endOffset = size is not an hidden invariant we depend on.

Reply:
I've added a test for the end offset.